homeassistant/components/coinbase/__init__.py

```diff
@@ -1,110 +1,150 @@
+            cv.deprecated(CONF_API_KEY),
             {
                 vol.Required(CONF_API_KEY): cv.string,
-                vol.Required(CONF_API_SECRET): cv.string,
-                vol.Optional(CONF_ACCOUNT_CURRENCIES): vol.All(
-                    cv.ensure_list, [cv.string]
-                ),
-                vol.Optional(CONF_EXCHANGE_CURRENCIES, default=[]): vol.All(
+                vol.Required(CONF_YAML_API_TOKEN): cv.string,
+                vol.Optional(CONF_CURRENCIES): vol.All(cv.ensure_list, [cv.string]),
+                vol.Optional(CONF_EXCHANGE_RATES, default=[]): vol.All(
                     cv.ensure_list, [cv.string]
                 ),
-            }
+            },
         )
     },
     extra=vol.ALLOW_EXTRA,
 )
 
 
-def setup(hass, config):
-    """Set up the Coinbase component.
-
-    Will automatically setup sensors to support
-    wallets discovered on the network.
-    """
-    api_key = config[DOMAIN][CONF_API_KEY]
-    api_secret = config[DOMAIN][CONF_API_SECRET]
-    account_currencies = config[DOMAIN].get(CONF_ACCOUNT_CURRENCIES)
-    exchange_currencies = config[DOMAIN][CONF_EXCHANGE_CURRENCIES]
-
-    hass.data[DATA_COINBASE] = coinbase_data = CoinbaseData(api_key, api_secret)
-
-    if not hasattr(coinbase_data, "accounts"):
-        return False
-    for account in coinbase_data.accounts:
-        if account_currencies is None or account.currency in account_currencies:
-            load_platform(hass, "sensor", DOMAIN, {"account": account}, config)
-    for currency in exchange_currencies:
-        if currency not in coinbase_data.exchange_rates.rates:
-            _LOGGER.warning("Currency %s not found", currency)
-            continue
-        native = coinbase_data.exchange_rates.currency
-        load_platform(
-            hass,
-            "sensor",
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Coinbase component."""
+    if DOMAIN not in config:
+        return True
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
             DOMAIN,
-            {"native_currency": native, "exchange_currency": currency},
-            config,
+            context={"source": SOURCE_IMPORT},
+            data=config[DOMAIN],
+        )
+    )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Coinbase from a config entry."""
+
+    # Do a one-time handle of options from YAML
+    if entry.source == SOURCE_IMPORT and not entry.options:
+        import_options: dict[str, list[str]] = {
+            CONF_CURRENCIES: [],
+            CONF_EXCHANGE_RATES: [],
+        }
+        if CONF_CURRENCIES in entry.data:
+            import_options[CONF_CURRENCIES] = entry.data[CONF_CURRENCIES]
+        if CONF_EXCHANGE_RATES in entry.data:
+            import_options[CONF_EXCHANGE_RATES] = entry.data[CONF_EXCHANGE_RATES]
+
+        hass.config_entries.async_update_entry(entry, options=import_options)
```

Comment:
This can be handled by the config flow when creating the entry.
The entry creation can, besides `data`, have initial `options` too.

That keeps the import handling nice and localized.

e.g., `self.async_create_entry(title="", data=user_input, options=imported_option_thingies)`

Reply:
Much better thanks :) I bottled actually using "thingies" though 🤣